src/Functions/punycode.cpp

```diff
@@ -0,0 +1,440 @@
+#include "config.h"
+
+#if USE_IDNA
+
+#include <Columns/ColumnNullable.h>
+#include <Columns/ColumnString.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypeString.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+
+#ifdef __clang__
+#    pragma clang diagnostic push
+#    pragma clang diagnostic ignored "-Wnewline-eof"
+#endif
+#    include <ada/idna/punycode.h>
+#    include <ada/idna/to_ascii.h>
+#    include <ada/idna/to_unicode.h>
+#    include <ada/idna/unicode_transcoding.h>
+#ifdef __clang__
+#    pragma clang diagnostic pop
+#endif
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int ILLEGAL_COLUMN;
+}
+
+namespace
+{
+
+enum class ErrorHandling
+{
+    Throw,
+    Null
+};
+
+/// Implementation of
+/// - punycode(En|De)code[OrNull](), see [1]
+/// - idna(En|De)code[OrNull](), see [2, 3]
```

Comment:
I understand about `punycodeDecode` but how `punycodeEncode` can fail? I mean why do we need `punycodeEncodeOrNull`?

Reply:
As far as I see, the only way for punycode encoding to fail currently is the (weird) edge that the input is not UTF-8. But from a ClickHouse POV, the ada-idna library API simply returns "failed" and there could be further sources of errors in future (e.g. a failure to allocate memory). For these reasons, I guess we should keep `punycodeEncodeOrNull()`. I can make the docs more concise tomorrow.

Reply:
From looking at the code of ada-idna it seems right now it returns "failed" statuses only when there are some coding errors, and not on allocation failures. If it can't allocate it will probably throw `std::bad_alloc`.

We don't usually handle too much if the input is not a proper UTF-8 string, see [upperUTF8](https://clickhouse.com/docs/en/sql-reference/functions/string-functions#upperutf8) or [base64Encode](https://clickhouse.com/docs/en/sql-reference/functions/string-functions#base64encode).

Reply:
Okay, I changed the code so that these functions remain:

- `punycodeEncode()`, `punycodeDecode()` and `tryPunycodeDecode()`
- `idnaEncode()`, `tryIdnaEncode()` and `idnaDecode()`

This is more consistent with other SQL functions.

Yes, it is still a bit unexpected why in the case of `idna` encoding may fail but this is required by the algorithm and I'll document it.

Reply:
Shouldn't it be `idnaEncode()`, `tryIdnaDecode()` and `idnaDecode()`?

Reply:
Oh, I see. It will probably also fail for long domains, for example
```
idnaEncode('1234567890123456789012345678901234567890123456789012345678901234')
```